File: app/src/main/java/com/josephlimbert/weighttracker/view/SetGoalWeightFragment.java
package com.josephlimbert.weighttracker.view;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.josephlimbert.weighttracker.R;
import com.josephlimbert.weighttracker.model.Weight;
import com.josephlimbert.weighttracker.viewmodel.UserViewModel;
import com.josephlimbert.weighttracker.viewmodel.WeightViewModel;

public class SetGoalWeightFragment extends BottomSheetDialogFragment {
    EditText weightText;
    WeightViewModel weightViewModel;
    float goalWeight;

    @NonNull
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View rootView =  inflater.inflate(R.layout.sheet_set_goal_weight, container, false);
        // Initialize variables
        weightText = rootView.findViewById(R.id.goal_weight_input_edit_text);
        TextView goalWeightLabel = rootView.findViewById(R.id.set_goal_label);
        Button submitButton = rootView.findViewById(R.id.set_goal_weight_submit_button);
        weightViewModel = new ViewModelProvider(requireActivity()).get(WeightViewModel.class);

        // Check for boolean argument indicating whether this view was created for editing
        // an existing goal weight or setting a new one.
        if (getArguments() != null) {
            // Store existing goal weight if we are editing a goal weight
            if (getArguments().getBoolean("isEditing")) {
                if (weightViewModel.getGoalWeight().getValue() != null) {
                    goalWeight = weightViewModel.getGoalWeight().getValue();
                    weightText.setText(String.valueOf(goalWeight));
                    goalWeightLabel.setText(getString(R.string.change_goal_weight));
                }
            }
        }

        submitButton.setOnClickListener(this::submitWeight);
        return rootView;
    }

    // This function will submit the goal weight to the database
    public void submitWeight(View v) {
        // Check that the goal weight is not empty
        if (weightText.getText().toString().isEmpty()){
            weightText.setError(getString(R.string.empty_weight_error));
        }
        else {
            try {
                float newWeight = Float.parseFloat(weightText.getText().toString());
                // If we are editing a goal weight then we update the weight and send the update to the database
                // Otherwise, we create a new goal weight and add it to the database
                if (weightViewModel.getGoalWeight().getValue() != null) {
                    goalWeight = newWeight;
                    weightViewModel.updateGoalWeight(goalWeight);
                    dismiss();
                } else {
                    weightViewModel.addGoalWeight(newWeight);
                    dismiss();
                }
            } catch (NumberFormatException e) {
                weightText.setError(getString(R.string.weight_number_error));
            }
        }
    }
}

File: app/src/main/java/com/josephlimbert/weighttracker/ui/sheet/SetGoalWeightFragment.kt
package com.josephlimbert.weighttracker.ui.sheet

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.josephlimbert.weighttracker.R
import com.josephlimbert.weighttracker.data.repository.FirestoreResult
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class SetGoalWeightFragment : BottomSheetDialogFragment() {
    var weightText: EditText? = null
    private val viewModel: SetGoalViewModel by viewModels()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        val rootView = inflater.inflate(R.layout.sheet_set_goal_weight, container, false)
        // Initialize variables
        weightText = rootView.findViewById(R.id.goal_weight_input_edit_text)
        val goalWeightLabel = rootView.findViewById<TextView>(R.id.set_goal_label)
        val submitButton = rootView.findViewById<Button>(R.id.set_goal_weight_submit_button)

        var userId: String? = null

        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.user.collect { user ->
                if (user != null) {
                    userId = user.id
                    if (arguments != null) {
                        if (requireArguments().getBoolean("isEditing")) {
                                weightText!!.setText(user.goalWeight.toString())
                                goalWeightLabel.text = getString(R.string.change_goal_weight)
                        }
                    }
                }
            }
        }

        // Check for boolean argument indicating whether this view was created for editing
        // an existing goal weight or setting a new one.


        submitButton.setOnClickListener { _: View? -> this.submitWeight(userId) }
        return rootView
    }

    // This function will submit the goal weight to the database
    fun submitWeight(userId: String?) {
        // Check that the goal weight is not empty
        if (weightText!!.text.toString().isEmpty()) {
            weightText!!.error = getString(R.string.empty_weight_error)
        } else {
            try {
                val newWeight = weightText!!.text.toString().toDouble()
                // If we are editing a goal weight then we update the weight and send the update to the database
                // Otherwise, we create a new goal weight and add it to the database
                viewLifecycleOwner.lifecycleScope.launch {
                    if (userId == null) return@launch
                    when (val result = viewModel.setGoalWeight(userId, newWeight)) {
                        is FirestoreResult.Success -> {
                            dismiss()
                        }
                        is FirestoreResult.Error -> {
                            Toast.makeText(context, result.message, Toast.LENGTH_LONG).show()
                        }
                    }
                }
            } catch (_: NumberFormatException) {
                weightText!!.error = getString(R.string.weight_number_error)
            }
        }
    }
}